File: luatest/Assets/StreamingAssets/lua/Robot.lua
function OnUpdate_Robot(robot, deltaTime) 
	

	state = robot.state
	if (state == nil) then
		state = "idle"
		

	end

	if (state == "reset") then
		robot:DropAll()
		state = "idle"
		robot:ReturnWork()
		

	elseif (state == "idle") then
		findWorkTimer = robot.info:GetFloat("findWorkTimer");
		if (findWorkTimer <= 0) then
			state = "find_work"
			findWorkTimer = 1
			robot.info:SetFloat("waitForWorkTimer", 2)
		else	
			findWorkTimer = findWorkTimer - deltaTime
		end
		robot.info:SetFloat("findWorkTimer", findWorkTimer)

		wanderTimer = robot.info:GetFloat("wanderTimer", 5)
		if (wanderTimer <= 0) then
			wanderTimer = 5
			state = "wander"
			robot.info:SetString("stateWhenMoved", "idle")
		else 
			wanderTimer = wanderTimer - deltaTime
			
		end
		robot.info:SetFloat("wanderTimer", wanderTimer)

	elseif  (state == "find_work") then
		waitForWorkTimer = robot.info:GetFloat("waitForWorkTimer")
		
		if (waitForWorkTimer <= 0) then
			state = "idle"
		else
			waitForWorkTimer = waitForWorkTimer - deltaTime
		end
		robot.info:SetFloat("waitForWorkTimer", waitForWorkTimer)
	--[[
		robot:GetWork()
		if (robot.work ~= nil) then
			state = "init_work"
		else	
			state = "idle"
		end
		]]
	elseif (state == "init_work") then
		if (robot.work ~= nil) then
			if (robot.work.inventoryItemName ~= nil) then
				state = "find_item"
			else
				--go to work tile
				state = "find_path_work.tile"
				robot.info:SetString("stateWhenMoved", "work")
			end
		else
			state = "idle"
		end
	elseif (state == "wander") then
		target = World.current:GetRandomEmptyTile()
		if (target ~= nil) then
			if (robot:FindPath(target,false)) then
				state = "move"	
			
			end
		else
			state = "idle"
		end
	elseif (state == "find_item") then
		--find some of that then
		target = World.current.inventoryManager:GetNearest(robot.pos, robot.work.inventoryItemName,robot.work.inventorySearchStockpiles)
		--World.dbug("target:" + target.ToString() + "\nitem:" + work.inventoryItemName)
		if (target ~= nil) then
			robot.info:SetInt("goto_x", target.world_x);		
			robot.info:SetInt("goto_y", target.world_y);		
			state = "find_path_x_y"
			robot.info:SetString("stateWhenMoved", "pickup")
			qtyToAllocate = robot.work.inventoryItemQtyRemaining - robot.inventory:HowMany(robot.work.inventoryItemName)
			target:InventoryAllocate(robot.work.inventoryItemName, qtyToAllocate)


		end		
	elseif (state == "pickup") then
		fx = robot.info:GetInt("goto_x")
		fy = robot.info:GetInt("goto_y")
		tile = World.current:GetTileAt(fx, fy)
		qtyToTake = robot.work.inventoryItemQtyRemaining - robot.inventory:HowMany(robot.work.inventoryItemName)


		if (robot:Pickup(tile, robot.work.inventoryItemName, qtyToTake)) then
			if (robot.inventory:HowMany(robot.work.inventoryItemName) >= robot.work.inventoryItemQtyRequired) then
				state = "find_path_work.tile"
				robot.info:SetString("stateWhenMoved", "work")
			else
				state = "find_item"
			end
		else
			state = "find_item"
		end
	elseif (state == "work") then
		robot.work:Work(deltaTime)

	elseif (state == "find_path_x_y") then
		fx = robot.info:GetInt("goto_x")
		fy = robot.info:GetInt("goto_y")

		if (robot:FindPath(fx, fy, true)) then
			state = "move"
		end
	elseif (state == "find_path_work.tile") then
		if (robot:FindPath(robot.work.workTile,true)) then
			state = "move"
		else	
			state = "reset"
		end
	elseif (state == "move") then
		if (robot.pos == robot.dst) then

			dst = robot.path:GetNextTile()
			if (dst == nil) then
				robot.path = nil
				state = "finished_moving"
			else	
				robot:SetDst(dst)

			end
		else
			robot:Move(deltaTime)
			
		end
	elseif (state == "finished_moving") then
		stateWhenMoved = robot.info:GetString("stateWhenMoved")
		if (stateWhenMoved == nil or stateWhenMoved == "") then
			state = "idle"
		else		
			state = stateWhenMoved
		end
		robot.info:SetString("stateWhenMoved", "idle")
	end

	robot.state = state
	
	
	sayTimer = robot.info:GetFloat("sayTimer")
	if (sayTimer <= 0) then
		robot:Say(state)
		sayTimer = 2
	else 
		sayTimer = sayTimer - deltaTime
	end
	robot.info:SetFloat("sayTimer", sayTimer)


	
	--[[
	
	
	if (robot.pos == robot.dst) then
		t = robot.pos:GetRandomNeighbour()
		if (t ~= nil) then
			robot:SetDst(t)
			robot:Move(deltaTime)
		else
			
		end 
	else
		robot:Move(deltaTime)
	end
	]]

end

--[[

	if nearby job requires materials
		fetch materials

	if nearby stockpile is not full
		fill

	if nearby stockpile is completely empty
		fill

	if nearby tile needs work
		work tile

	fetch materials
		find materials
		foreach 
			allocate
			add to list
		
		foreach
			find path
			go
			pick up
			if full goto end


]]

File: luatest/Assets/Scripts/DataModels/Robot.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using UnityEngine;

namespace NoYouDoIt.DataModels {
  using NoYouDoIt.TheWorld;
  using NoYouDoIt.NYDIPaths;
  using NoYouDoIt.Utils;
  using NoYouDoIt.Controller;
  using System.IO;
  using Newtonsoft.Json.Linq;

  public class Robot {

    Action<Robot> CBOnChanged;

    public TileOccupant occupier;
    private float waitTimer;
    public string state;
    public string NewState = null;
    public string typeName { get; private set; }
    public Robot proto { get; private set; }

    public ItemParameters info;
    public string name { get; private set; }
    public readonly string uuid = Guid.NewGuid().ToString();
    public Tile pos { get; private set; }
    public Tile dst { get; private set; }
    private float moveProgress = 0;

    private Vector2 vPos = new Vector2();


    //PathAS currentPath;


    public Tile target { get; private set; } = null;
    public TilePathAStar path;
    string currentInstruction;
    public WorkItem work;
    public Inventory inventory;
    public string spriteName { get; private set; }
    string OnUpdate = "OnUpdate_Robot";

    public int Xint {
      get {
        return pos.world_x;
      }
    }

    public int Yint {
      get {
        return pos.world_y;
      }
    }
    public float X {
      get {
        return Mathf.Lerp(pos.world_x, dst.world_x, moveProgress);
      }
    }
    public float Y {
      get {
        return Mathf.Lerp(pos.world_y, dst.world_y, moveProgress);
      }
    }
    public Vector2 position {
      get {
        vPos.Set(X, Y);
        return vPos;
      }
    }

    public void Say(string s) {
      WorldController.Instance.SpawnText(s, Xint, Yint);
    }



    public void PleaseMove() {
      if (state == "idle") {
        NewState = "wander";
        state = "wander";
      }
    }


    private Robot() {
      
    }
    private Robot(Tile tile, Robot proto) {
      inventory = new Inventory(World.current, 1, INVENTORY_TYPE.ROBOT, this);
      name = World.current.GetName();
      info = new ItemParameters(proto.info);
      info.SetFloat("movement_speed", 5);
      this.typeName = proto.typeName;
      this.spriteName = proto.spriteName;
      this.proto = proto;
      this.SetPos(tile);
      this.SetDst(tile);
      occupier = new TileOccupant(name, proto.typeName);
      occupier.CBPleaseMove += PleaseMove;



    }

    public void ReturnWork() {
      World.current.workManager.ReturnWork(work);
    }

    private void DropAll() {

      while (!inventory.IsEmpty()) {
        Tile t = World.current.FindEmptyTile(pos);
        if (t != null) {
          string itemName = inventory.GetFirst();
          int qtyToPlace = inventory.HowMany(itemName);
          int qtyPlaced = t.AddToInventory(itemName, qtyToPlace);
          inventory.RemoveItem(itemName, qtyPlaced);
        } else {
          break;
        }
      }
    }

    public static Robot MakeRobot(Tile t, string typeName) {



      if (t.IsEmpty() && t.countOfOccupied == 0) {
        Robot proto = GetPrototype(typeName);
        if (proto != null) {
          Robot robot = new Robot(t, proto);
          robot.pos = t;
          robot.dst = t;

          return robot;
        } else {
          Debug.LogError("could not find robot type of name [" + typeName + "]");
        }

        return null;
      }

      return null;

    }



    public bool GiveInstruction(string instruction) {

      if (currentInstruction == null) {
        currentInstruction = instruction;
        return true;
      } else {
        return false;
      }
    }

    public void Update(float deltaTime) {
      float old = (X * 11) + (Y * 3);
      World.CallLuaFunction(OnUpdate, this, deltaTime);

      float n = (X * 11) + (Y * 3);

      if (old != n) {
        if (CBOnChanged != null) {
          CBOnChanged(this);
        }
      }

    }

    public void Work(float deltaTime) {
      World.CallLuaFunctions(work.OnWork.ToArray(), this, deltaTime);
    }

    public bool PlaceItemAtJob() {
      if (work != null) {
        string holding = work.inventoryItemName;
        int holdingQty = inventory.HowMany(holding);
        int placed = work.inventory.AddItem(holding, holdingQty);
        inventory.RemoveItem(holding, placed);
        work.inventoryItemQtyRemaining -= placed;
        //Debug.Log("robot.PlaceItemAtJob: " + work);
        return true;
      }
      return false;
    }

    public bool PlaceItemOnTile() {
      if (work != null) {
        string holding = work.inventoryItemName;
        int holdingQty = inventory.HowMany(holding);
        int placed = work.workTile.AddToInventory(holding, holdingQty);
        inventory.RemoveItem(holding, placed);
        work.inventoryItemQtyRemaining -= placed;
      }

      return false;
    }

    public bool PlaceItem(Inventory inventory, string itemName, int qty) {



      return false;
    }

    public bool Pickup(Tile t, string itemName, int qty) {

      if (t == pos || t.IsNeighbour(pos, false)) {
        int qtyTaken = t.RemoveFromInventory(itemName, qty);
        //Debug.Log("qty actually removed:" + qtyTaken);
        if (qtyTaken == 0) return false;

        int qtyAdded = inventory.AddItem(itemName, qtyTaken);
        if (qtyAdded < qtyTaken) {
          t.AddToInventory(itemName, qtyTaken - qtyAdded);

        }
        return true;

      }

      return false;



    }

    public void Move(float deltaTime) {
      if (pos.movementFactor == 0 || dst.movementFactor == 0) {
        NewState = "find_new_path";
        SetPos(World.current.FindEmptyTile_NotThisOne(pos));
        SetDst(pos);

      }


      if (pos != dst) {
        float distToTravel = Funcs.Distance(pos, dst);
        bool never = false, soon = false;
        switch (dst.CanEnter(occupier)) {
          case Tile.CAN_ENTER.YES:
            break;
          case Tile.CAN_ENTER.NEVER:
            never = true;
            break;
          case Tile.CAN_ENTER.SOON:
            soon = true;
            break;
          default:
            break;
        }

        if (never) {
          NewState = "find_new_path";
        } else if (soon) {
          //wait
          dst.PleaseMove();
          waitTimer += deltaTime;

          if (waitTimer > 3) {
            NewState = "find_new_path";
            waitTimer = 0;
          }

        } else {


          float speed = info.GetFloat("movement_speed");

          speed *= Mathf.Clamp(dst.movementFactor, 0.2f, 1.5f);

          float distThisFrame = speed * deltaTime;

          moveProgress += (distThisFrame / distToTravel);

          if (moveProgress >= 1) {
            SetPos(dst);
            moveProgress = 0;
          }
        }
      } else {
        moveProgress = 0;
      }
    }

    public bool SetDst(Tile tile) {
      if (tile != null && tile.movementFactor > 0 && (tile.IsNeighbour(pos, true) || tile == pos)) {
        dst = tile;
        return true;
      } else {
        //Debug.Log("Character SetDestination was given a tile that was not a neighbour of the current tile...");
        return false;
      }



    }

    public bool SetPos(Tile t) {
      if (t != pos) {
        if (pos != null) {
          //leave the tile
          pos.Leave(occupier);

        }
        pos = t;
        pos.Enter(occupier);
        return true;


      }
      return false;

    }

    public bool FindPath(int x, int y, bool neighboursOk) {
      Tile end = World.current.GetTileAt(x, y);
      if (end == null) return false;

      path = PathFinder.FindPath(World.current, pos, end, neighboursOk, neighboursOk);
      if (path != null && path.foundPath) {
        return true;
      } else {
        path = null;
        return false;
      }
    }

    public bool FindPath(Tile end, bool neighboursOk) {

      path = PathFinder.FindPath(World.current, pos, end, neighboursOk, neighboursOk);
      if (path != null && path.foundPath) {
        return true;
      } else {
        path = null;
        return false;
      }

    }


    public void CBRegisterOnChanged(Action<Robot> cb) {
      CBOnChanged += cb;

    }

    public void CBUnregisterOnChanged(Action<Robot> cb) {
      CBOnChanged -= cb;
    }
    //-------------------------------- STATIC- -------------------------------------
    private static Dictionary<string, Robot> prototypes = new Dictionary<string, Robot>();

    public static void LoadFromFile() {
      string path = Path.Combine(Application.streamingAssetsPath, "data", "RobotTypes");

      string[] files = Directory.GetFiles(path, "*.json");

      foreach (string file in files) {
        string fcontents = File.ReadAllText(file);
        JObject json = JObject.Parse(fcontents);

        CreateRobotPrototype(json);
      }
    }

    private static void CreateRobotPrototype(JObject json) {
      Robot proto = new Robot();
      proto.name = "prototype";
      proto.typeName = Funcs.jsonGetString(json["name"], null);
      proto.spriteName = Funcs.jsonGetString(json["spriteName"], null);
      proto.info = new ItemParameters();

      prototypes.Add(proto.typeName, proto);
    }

    private static Robot GetPrototype(string typeName) {
      if (prototypes.ContainsKey(typeName)) {
        return prototypes[typeName];
      } else {
        return null;
      }
    }
  }





}
